pkg/grpc/tls_client_certificate_authenticator.go

```diff
@@ -61,5 +65,49 @@ func (a *tlsClientCertificateAuthenticator) Authenticate(ctx context.Context) (*
 	if _, err := certs[0].Verify(opts); err != nil {
 		return nil, util.StatusWrapWithCode(err, codes.Unauthenticated, "Cannot validate TLS client certificate")
 	}
-	return a.metadata, nil
+
+	searchContext, err := getClientCertificateJMESPathSearchContext(certs[0])
+	if err != nil {
+		return nil, util.StatusWrapWithCode(err, codes.Unauthenticated, "Failed to create a search context for extracting metadata")
+	}
+
+	// Validate the client cert matches our expectations
+	validationResult, err := a.validator.Search(searchContext)
+	if err != nil {
+		return nil, util.StatusWrapWithCode(err, codes.Unauthenticated, "TLS client certificate failed JMESPath validation")
+	}
+	if validationResult != true {
+		return nil, util.StatusWrapWithCode(errors.New("JMESPath validation result was not true"), codes.Unauthenticated, "TLS client certificate failed JMESPath validation")
```

Comment:
You can just return `status.Error(codes.Unauthenticated, "....")` here.

Reply:
Also please make sure that these error messages don't contain too many technicalities. Keep in mind that these error messages are returned to the user. The JMESPath expressions are something that an administrator of the cluster specifies; not the user.